felix/nftables/maps.go

```diff
@@ -0,0 +1,677 @@
+	defer func() {
+		s.logCxt.WithFields(logrus.Fields{
+			"resyncDuration":         time.Since(resyncStart),
+			"mapsWithDirtyMembers":   s.mapsWithDirtyMembers.Len(),
+			"mapsToCreateOrRecreate": s.mapNameToProgrammedMetadata.PendingUpdates().Len(),
+			"mapsToDelete":           s.mapNameToProgrammedMetadata.PendingDeletions().Len(),
+		}).Debug("Finished Maps resync")
+	}()
+
+	// Clear the dataplane metadata view, we'll build it back up again as we scan.
+	s.mapNameToProgrammedMetadata.Dataplane().DeleteAll()
+
+	// Load from the dataplane. Update our Dataplane() maps with the actual contents
+	// of the data plane.
+	//
+	// For any map that doesn't match the desired data plane state, we'll queue up an update.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	maps, err := s.nft.List(ctx, "map")
+	if err != nil {
+		if knftables.IsNotFound(err) {
+			// Table doesn't exist - nothing to resync.
+			return nil
+		}
+		return fmt.Errorf("error listing nftables maps: %s", err)
+	}
+
+	// We'll process each map in parallel, so we need a struct to hold the results.
+	// Once knftables is augmented to support reading many maps at once, we can remove this.
+	type mapData struct {
+		name  string
+		elems []*knftables.Element
+		err   error
+	}
+	mapsCh := make(chan mapData)
+	defer close(mapsCh)
+
+	// Start a goroutine to list the elements of each map. Limit concurrent map reads to
+	// avoid spawning too many goroutines if there are a large number of maps.
+	routineLimit := make(chan struct{}, 100)
+	defer close(routineLimit)
+	for _, name := range maps {
+		// Wait for room in the limiting channel.
+		routineLimit <- struct{}{}
+
+		// Start a goroutine to read this map.
+		go func(name string) {
+			// Make sure to indicate that we're done by removing ourselves from the limiter channel.
+			defer func() { <-routineLimit }()
+
+			elems, err := s.nft.ListElements(ctx, "map", name)
+			if err != nil {
+				mapsCh <- mapData{name: name, err: err}
+				return
+			}
+			mapsCh <- mapData{name: name, elems: elems}
+		}(name)
+	}
+
+	// We expect a response for every map we asked for.
+	responses := make([]mapData, len(maps))
+	for i := range responses {
+		mapData := <-mapsCh
+		responses[i] = mapData
+	}
+
+	for _, mapData := range responses {
+		mapName := mapData.name
+		logCxt := s.logCxt.WithField("mapName", mapName)
+		if mapData.err != nil {
+			logCxt.WithError(err).Error("Failed to list map elements.")
+			return mapData.err
+		}
+
+		// TODO: We need to be able to extract the map type from the dataplane, otherwise we cannot
+		// tell whether or not the map has the correct type.
+		metadata, ok := s.mapNameToAllMetadata[mapName]
+		if !ok {
+			// Programmed in the data plane, but not in memory. We should still load any members of this map in order
+			// to perform map deletion logic (delete members, delete map).
+			logCxt.Info("Map in dataplane but not in memory, will remove it.")
```

Comment:
Was expecting to see you add something to `mapNameToProgrammedMetadata.Dataplane()` to tell the later logic that there's something we're not expecting.